File: DataStructure/Union_Find/q9_591-Connecting-Graph-iii.py
class ConnectingGraph3:
    def __init__(self, n):
        self.father = {}
        self.cnt_connected = n  # 连通块个数
        for i in range(n + 1):
            self.father[i] = i

    def connect(self, a, b):
        root_a, root_b = self.find(a), self.find(b)
        if root_a != root_b:
            self.father[root_a] = root_b
            self.cnt_connected -= 1

    def query(self):
        return self.cnt_connected

    def find(self, x):
        if self.father[x] == x:
            return x
        self.father[x] = self.find(self.father[x])
        return self.father[x]

File: DataStructure/Union_Find/q9_591_Connecting_Graph_iii.java
package DataStructure.Union_Find;

public class q9_591_Connecting_Graph_iii {
    private int[] father;
    private int connCnt;
    public q9_591_Connecting_Graph_iii(int n) {
        father = new int[n+1];
        connCnt = n;
        for (int i = 0; i <= n; i++)
            father[i] = i;

    }

    public void connect(int a, int b) {
        int rootA = findR(a), rootB = findR(b);
        if (rootA != rootB) {
            father[rootA] = rootB; // 必须是rootA/B，而非a/b!!!
            connCnt--;
        }
    }

    public int query() {
        return connCnt;
    }

    private int findR(int x) {
        if (father[x] != x) {
            father[x] = findR(father[x]);
        }
        return father[x];
    }
}
